File: python/114_Flatten_Binary_Tree_to_Linked_List.py
#!/usr/bin/env python3

import TreeMethod

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if root is None:
            return None

        cur = root
        stack = []
        stackR = []
        point = TreeNode(0)

        while len(stack) > 0 or cur:
            while cur:
                point.right = cur
                point = cur
                tmp = cur
                stack.append(cur)
                cur = cur.left
                tmp.left = None
                stackR.append(tmp.right)

            stack.pop()
            cur = stackR.pop()
        return root

s = Solution()
print(TreeMethod.show(s.flatten(TreeMethod.build([0]))))

File: java/src/brand_new/_114_Flatten_Binary_Tree_to_Linked_List.java
package brand_new;

import java.util.Stack;

public class _114_Flatten_Binary_Tree_to_Linked_List {
	public void flatten(TreeNode root) {
		Stack<TreeNode> stack = new Stack<>();
		TreeNode cur = root;

		while (cur != null || stack.size() > 0) {
			while (cur.left != null) {
				if (cur.right != null)
					stack.push(cur.right);
				cur.right = cur.left;
				cur.left = null;
				cur = cur.right;
			}

			if (cur.right != null)
				stack.push(cur.right);

			if (stack.size() > 0) {
				TreeNode right = stack.pop();
				cur.right = right;
				cur = cur.right;
			} else
				break;
		}
	}

	public static void main(String[] args) {

	}
}
